developer_guides/virtualisation/virtualisation.rst

```diff
@@ -0,0 +1,255 @@
+guest pipeline on the host requires changes to the ASoC core, without which the
+guest audio path is not recognised as inactive and is not pruned, thus making
+further streaming by the guest impossible.
+
+Upon reception of the STREAM_TRIG_START IPC message the vhost driver activates
+the virtual PCM pipeline and updates the routing information on the host.
+
+As mentioned above, it is our goal to implement data zero-copy. As long as
+this isn't the case, the guest VirtIO driver implements .block_write() and
+.block_read() SOF DSP operation methods for data streaming via the data
+VirtQueue.
+
+The host receives requests on the data VirtQueue and copies data between PCM
+runtime buffers and the queue and responds either with a status or with a data
+buffer.
+
+Similarly to the data, streaming buffer position updates are currently also
+transferred via a dedicated VirtQueue. Both the host and the guest are
+configured to use IPC messages for position updates. Buffers for VirtIO
+VirtQueues are always provided by guests, therefore there should always be a
+position update buffer waiting on the host side to send a position update
+message as soon as one arrives from the DSP.
+
+Upon reception of the STREAM_TRIG_STOP IPC message the vhost driver updates
+the routing information and deactivates the virtual PCM pipeline on the host.
+This operation requires particular care, because in the present state, just
+calling
+
+	soc_dpcm_runtime_update();
+	snd_soc_runtime_deactivate();
+
+doesn't deactivate the pipeline and therefore doesn't allow runtime PM to
+suspend the interface, which then breaks following activation attempts. The
+current VirtIO implementation contains a fix for that, but that fix might turn
+out to be incorrect. The difficulty in establishing a correct fix is, that so
+far it hasn't been possible to find a test case for
+snd_soc_runtime_deactivate().
+
+3.3. Deactivation
+-----------------
+
+When the guest isn't actively using audio, we want to allow the host to
+runtime suspend the DSP. That usually means switching the DSP off. Next time the
+guest wants to activate audio it will have to perform initialisation again,
+including sending its topology to the host. This happens automatically as long
+as the guest runtime suspends too or performs the initialisation cycle in some
+other way, by unloading and reloading drivers, or by rebooting.
+
+With automatic runtime suspending on the host and the guest, it can happen,
+that one of them got suspended, whereas the other one is still running. A
+problematic case would be, if the host got suspended, while the guest was still
+active, because then, when the host resumes, it will also bring up the DSP and
+only send the host topology part to it. The guest however wasn't suspended,
+therefore it won't send its topology to the host. To avoid this the automatic
+runtime suspend timeout on the guest is set to a half of the host's timeout.
```

Comment:
this is really confusing. pm_runtime is all about devices, and here we have vhost/guest devices. I really have no idea how pm_runtime can be synchronized unless there's a clear protocol and refcounting. 

BTW why this would this be audio specific, we can't be the only ones enabling pm_runtime?

Reply:
ack, ref counting on host for all guests will manage this in general purpose way.